### ISSUE TITLE
all the wait_for & wait_until are broken

### ISSUE THREAD

**Issue by sandym** [klemens-morgenstern/boost-process#112]
At least on macOS, possibly on other posix platforms.

The code for those all have this structure:

```

    int status;
   // ...
    do
    {
        ret = ::waitpid(p.pid, &status, WUNTRACED | WNOHANG);
      // ...
    } 
    while (((ret == -1) && errno == EINTR)       || 
           ((ret != -1) && !WIFEXITED(status)));

```

The problem is that if the child is not finished, waitpid will return 0 and leave the status unchanged since there is no exit status to report. So the condition that will be evaluated in the while() is "(ret != -1) && !WIFEXITED(status)", here status is garbage, undefined behaviour, if you're lucky it might works.

Code to reproduce:

```
#include <boost/process.hpp>

void zero_stack()
{
    char stack[1000];
    memset( stack, 0, 1000 );
}

int main()
{
    auto c = boost::process::child( "/bin/sleep", "60" );
    //zero_stack(); <--- uncomment this line and it will fail for sure, on macOS
    c.wait_for( std::chrono::seconds(120) ); // this may or may not wait, try your luck!
}

```
And it's disappointing to see the parent process completely waste a cpu while waiting.

**Comment by Taredon** [comment]
I guess this is a duplicate of #99 . At least I see i'm not the only one affected by these issues on my linux systems.

**Comment by Taredon** [comment]
Since there seems no viable way to implement these methods correctly without having:
- some sleep() based loops
- to do some ugly pipe hacks which most likely wont work on windows

I guess the must suitable solutions would be going the same way as the [POCO Project](https://pocoproject.org/) and just remove them completely. So the user would still have the option to periodically check the process state via running().

**Comment by klemens-morgenstern** [comment]
Thanks, and thanks for being patient @Taredon.
Well, I'd actually think having it sleep inside a loop would be alright if it is documented and if the alternative with `asio` works. The other way would be some strange workaround or using `sigwaitinfo`.

I should however add, that it's a big mistake on my part, to not have proper tests for those functions.

**Comment by egorpugin** [comment]
There must be very small sleeps to act on died children really in time.

But maybe it's possible to create separate `io_service` in those functions, set `signal_set` handler, set timeout and just wait in kernel - `ios.run()` ?

**Comment by sandym** [comment]
@egorpugin, I was thinking along those line too. Qt has no problem doing it by the way: 

```
#include <QProcess>

int main( int argc, char **argv )
{
	auto p = new QProcess;
	p->setProgram( "/bin/sleep" );
	p->setArguments( { "300" } );
	p->start();
	p->waitForFinished( 600000 ); // this will wait, using no cpu
}

```

Which is what we use now, unfortunately, for the asynchronous notification we now need an extra thread to run a Qt event loop just to handle child processes on_exit, while the rest of the code is asio based...

**Comment by sandym** [comment]
@Taredon, sorry I read #99 right after I posted this one, yeah, it's exactly the same problem.

**Comment by Taredon** [comment]
Yeah #99 was somewhat dead since I didnt had any new ideas anyways. 
Good tip about Qt though. Maybe their implementation would be an option.

@klemens-morgenstern I Wasnt quite sure how to handle these methods. Doing sleep based loops could work, though im not sure how find an appropriate timing (since small intervals waste CPU and big intervals would waste (user) time if the childs are only short running processes (a call to /usr/bin/find for example could take less than a few milliseconds up to severals minutes)

**Comment by egorpugin** [comment]
It seems qt relies on pipes fds to track process liveness, so they wait for them with `poll`.
http://man7.org/linux/man-pages/man2/poll.2.html
For reference:
http://code.qt.io/cgit/qt/qt.git/plain/src/corelib/kernel/qcore_unix_p.h
http://code.qt.io/cgit/qt/qt.git/plain/src/corelib/kernel/qcore_unix.cpp

But is it possible to child application to close 0,1,2 file handles intentionally and continue to work.
In such case polling fds won't work anymore.

**Comment by klemens-morgenstern** [comment]
@egorpugin I don't think the wait has to be that small. The only thing that could happen is that the current child process dies and another one with the same PID appears. I find that very unlikely to happen within a few milliseconds.
I'll try to find a solutions with signals. You're right, we cannot use the I/O.

**Comment by egorpugin** [comment]
> The only thing that could happen is that the current child process dies and another one with the same PID appears.

No, it's impossible. Kernel still keeps zombie process until you execute your first `waitpid` call on it.

> You're right, we cannot use the I/O.

I hope you mean fd io, not io service.

**Comment by klemens-morgenstern** [comment]
1. Not impossible, if an io_service is listening; i.e. we can have several calls of `waitpid`. But that's so very, very unlikely.

2. Yes I meant FD.

**Comment by egorpugin** [comment]
If user won't mix async (io_sevice+on_exit) and sync approaches c.wait[_for/_until], everything is good.

**Comment by klemens-morgenstern** [comment]
Well that is possible, thus it should work at least with a very high probability

**Comment by sandym** [comment]
> But is it possible to child application to close 0,1,2 file handles intentionally and continue to work.

No, the poll wait on a pipe FROM the Qt signal handler, not from the child process. It's the way to transfer control from a signal handler to the main program flow. In boost-process case, asio is taking care of this (using the exact same technique by the way, a pipe from signal handler to asio event loop, see asio_signal_handler() in the source).

This would be the best way of doing this, wait on the first of a timer or the signal. Not sure about that, but I think you can have multiple handlers for a signal event in asio (?) so it could work even when mixing on_exit and sync method, would need to by careful about race conditions.

> ...it should work at least with a very high probability.

That's the spirit, works with high probability !

**Comment by egorpugin** [comment]
> No, the poll wait on a pipe FROM the Qt signal handler, not from the child process. ... same as asio ...

I see, thanks.

> would need to by careful about race conditions.

To workaround this I wrote my proposal earlier (https://github.com/klemens-morgenstern/boost-process/issues/109#issuecomment-317410501) to have *the single place* in code with `waitpid(2)` call.
Inside it we can put a simple mutex after member variable check. And our sync/async problem (race) will be resolved.

Now we have `waitpid(2)`s spilled on over many posix files. This causes many issues.

**Comment by sandym** [comment]
Yes, I liked that idea, single place for waitpid to record the exit status.

**Comment by klemens-morgenstern** [comment]
@egorpugin That does NOT work. If you wait using async you have to use waitpid to the exit code. That is especially true if the child gets detached or if you use async_system in the first place. The possibility to have several `waitpid` is given.
In addition, there should not be any logging, the exit_code is in a shared-pointer to an atomic int, which keeps that. I.e. whichever method catches that first stores it there.

**Comment by egorpugin** [comment]
>  If you wait using async you have to use waitpid to the exit code.

I'm confused. We're somewhere on different pages.
My statements do not contradict with this one and other yours on this topic. At the moment I do not see the issue you describe - because from my POV everything working now just as you describe.

(Of course it's possible that I'm getting something wrong.)

**Comment by klemens-morgenstern** [comment]
Well when `signal_set::async_wait` expires, the call-back does have to call `waitpid` to remove the process (which exited) from the table and get the exit-code. This does not keep the user from calling `wait` in another thread, thus there might be several places in the code where `waitpid` is called.

**Comment by egorpugin** [comment]
> the call-back does have to call waitpid to remove the process 

Yes.

> thus there might be several places in the code where waitpid is called. 

But I'm saying about internal wrapper for waitpid with some sync, e.g. based on current `_exit_status`.
One localized `waitpid` call site, one piece of any possible checks on `waitpid`'s return value, `errno` etc.

**Comment by klemens-morgenstern** [comment]
Alright, would [`ularam`](https://linux.die.net/man/3/ualarm) work? I.e. call it before `waitpid` and wake the thingy up?

That will not yet solve the `io-service` problem but at least the `wait_until`.

**Comment by klemens-morgenstern** [comment]
Looked it up, `boost::asio::signal_set` will not do the job. I'll need to figure out a way to share the results across several receivers.
Would you consider it an awful restriction, if all exit-codes waited for asynchronously required the usage of a single `io_service`?

**Comment by egorpugin** [comment]
@klemens-morgenstern, could you please summarize the list of related issues for obtaining process exit status?
I still think that my approach can solve all of them. Maybe I'll try implement and show/push it.

1. wait_for/wait_until not working
2. waitpid races
3 more?

**Comment by klemens-morgenstern** [comment]
1. `wait_for`/`wait_pid`
2. The async-wait (i.e. `on_exit` using `signal_set`) will not capture the correct process when several of those shall be captured.

**Comment by egorpugin** [comment]
>The async-wait (i.e. on_exit using signal_set) will not capture the correct process when several of those shall be captured.

When does this happen (`several of those`)? I thought I fixed this in my PRs.

**Comment by klemens-morgenstern** [comment]
Not really, travis fails all async_wait stuff.

**Comment by klemens-morgenstern** [comment]
Ok, I have a [compiling](https://travis-ci.org/klemens-morgenstern/boost-process/jobs/260082615) version [here](https://github.com/klemens-morgenstern/boost-process/tree/sigchld_service), which has the restriction, that you will have to use one `io_service` for all `signal(SIGCHLD)`. That does not mean that you have to use the same one for for the async-i/o though.

If you could try it out, I'd be super thankful, I'll probably will need to merge that today, because the master closes for bugfixes until 1.65 is released.

**Comment by keith-bennett-gbg** [comment]
Hi,

I created a ticket on boost trac, [here](https://svn.boost.org/trac10/ticket/13139) which I think is also related to this issue (or at the very least, the same code around this issue). I saw recently on the boost mailing list that this github repo is better for issues, so came here poking around and found this issue.

@klemens-morgenstern the [link](https://github.com/klemens-morgenstern/boost-process/tree/sigchld_service) you shared gives me a 404. I am going to assume that you've merged the branch to master and deleted the branch?

**Comment by klemens-morgenstern** [comment]
Oh sry, I completely missed that on trac. The issue was fixed (https://github.com/klemens-morgenstern/boost-process/blob/develop/include/boost/process/detail/posix/wait_for_exit.hpp#L46)

**Comment by klemens-morgenstern** [comment]
Is this still open?

**Comment by klemens-morgenstern** [comment]
Is this solved by 09381034276ce76eedd6936ae330159dec9938d9 ?

**Comment by klemens-morgenstern** [comment]
@sandym Hey Sandy, my fix using `sigtimedwait` now breaks the OSX build - do you have any idea how to solve the issue?
